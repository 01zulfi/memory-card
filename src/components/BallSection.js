import React, { useState } from "react";
import { balls } from "../Balls";

const BallSection = ({ score }) => {
  const [fourBalls, setFourBalls] = useState(balls.getFourRandom());

  const toggleFailureBorder = (element) => {
    element.style.border = "2px solid red";
  };

  const setBallClicked = (id, event) => {
    const targetBall = balls.find(id);

    if (targetBall.clicked) {
      score.failure();
      toggleFailureBorder(event.target);
      balls.setAllClickedFalse();
      setTimeout(() => {
        setFourBalls(balls.getFourRandom());
      }, 1000);
      return;
    }

    setFourBalls(balls.getFourRandom());
    score.success();
    targetBall.setClickedTrue();
  };

  const clickHandler = (event) => {
    const id = Number(event.target.getAttribute("data-id"));
    setBallClicked(id, event);
  };

  return (
    <section className="ball-section">
      {fourBalls.map((ball) => (
        <div
          data-id={ball.id}
          className="ball-div"
          onClick={clickHandler}
          key={ball.id}
        >
          {ball.name}
        </div>
      ))}
    </section>
  );
};

export default BallSection;
