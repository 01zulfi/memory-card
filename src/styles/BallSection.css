section.ball-section {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  width: 100%;
  margin-top: 15px;
  gap: 20px;
}

section.ball-section.clicked-twice {
  pointer-events: none;
}

div.ball-div {
  width: 45%;
  padding: 5px;
  background-color: rgb(182, 214, 204, 0.5);
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

div.ball-div.clicked-twice {
  border: 2px solid red;
  background-color: rgb(182, 214, 204, 0.1);
}

div.ball-div img {
  aspect-ratio: 1;
  width: 75%;
}

div.ball-div p {
  width: 100%;
  text-align: center;
  margin-top: 5px;
  margin-bottom: 5px;
  padding: 5px;
  font-size: 1.1rem;
  font-weight: 600;
  background-color: rgb(182, 214, 204, 0.8);
}
