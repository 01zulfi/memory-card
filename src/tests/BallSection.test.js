import React from "react";
import { screen, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import BallSection from "../components/BallSection";

test("renders ball section", () => {
  const { container } = render(<BallSection />);
  expect(container).toMatchSnapshot();
  expect(container.firstChild).toHaveClass("ball-section");
});

test("re-renders ball divs", () => {
  const { container } = render(<BallSection />);
  expect(container).toMatchSnapshot();
  screen.debug();

  const div = container.querySelector("div");
  userEvent.click(div);
  screen.debug();
  expect(container).toMatchSnapshot();
});
