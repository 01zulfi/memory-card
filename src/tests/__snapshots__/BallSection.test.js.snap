// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`re-renders ball divs 1`] = `
<div>
  <section
    class="ball-section"
  >
    <div>
      Beachball
    </div>
    <div>
      Ping Pong Ball
    </div>
    <div>
      Football
    </div>
    <div>
      Curling Ball
    </div>
  </section>
</div>
`;

exports[`re-renders ball divs 2`] = `
<div>
  <section
    class="ball-section"
  >
    <div>
      Cricket Ball
    </div>
    <div>
      Tennis Ball
    </div>
    <div>
      Bowling Ball
    </div>
    <div>
      Basketball
    </div>
  </section>
</div>
`;

exports[`renders ball section 1`] = `
<div>
  <section
    class="ball-section"
  >
    <div>
      Basketball
    </div>
    <div>
      American Football
    </div>
    <div>
      Beachball
    </div>
    <div>
      Ping Pong Ball
    </div>
  </section>
</div>
`;
